// @cucumber/cucumber/lib/formatter/builtin/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.documentation = void 0;
const json_formatter_1 = __importDefault(require("../json_formatter"));
const progress_formatter_1 = __importDefault(require("../progress_formatter"));
const progress_bar_formatter_1 = __importDefault(require("../progress_bar_formatter"));
const rerun_formatter_1 = __importDefault(require("../rerun_formatter"));
const snippets_formatter_1 = __importDefault(require("../snippets_formatter"));
const summary_formatter_1 = __importDefault(require("../summary_formatter"));
const usage_formatter_1 = __importDefault(require("../usage_formatter"));
const usage_json_formatter_1 = __importDefault(require("../usage_json_formatter"));
const message_1 = __importDefault(require("./message"));
const html_1 = __importDefault(require("./html"));
const builtin = {
    // new plugin-based formatters
    html: html_1.default,
    junit: '@cucumber/junit-xml-formatter',
    message: message_1.default,
    // legacy class-based formatters
    json: json_formatter_1.default,
    progress: progress_formatter_1.default,
    'progress-bar': progress_bar_formatter_1.default,
    rerun: rerun_formatter_1.default,
    snippets: snippets_formatter_1.default,
    summary: summary_formatter_1.default,
    usage: usage_formatter_1.default,
    'usage-json': usage_json_formatter_1.default,
};
exports.default = builtin;
exports.documentation = {
    // new plugin-based formatters
    html: 'Outputs a HTML report',
    junit: 'Produces a JUnit XML report',
    message: 'Emits Cucumber messages in newline-delimited JSON',
    // legacy class-based formatters
    json: json_formatter_1.default.documentation,
    progress: progress_formatter_1.default.documentation,
    'progress-bar': progress_bar_formatter_1.default.documentation,
    rerun: rerun_formatter_1.default.documentation,
    snippets: snippets_formatter_1.default.documentation,
    summary: summary_formatter_1.default.documentation,
    usage: usage_formatter_1.default.documentation,
    'usage-json': usage_json_formatter_1.default.documentation,
};
